torch/nn/modules/transformer.py

```diff
@@ -0,0 +1,519 @@
+        output = src
+        if self.src_embed:
+            output = self.src_embed(output)
+
+        if self.pos_encoder:
+            output = self.pos_encoder(output)
+
+        for i in range(self.num_layers):
+            output = self.layers[i](output, mask)
+
+        if self.norm:
+            output = self.norm(output)
+
+        return output
+
+
+class TransformerDecoder(Module):
+    r"""TransformerDecoder is a stack of N decoder layers
+
+    Args:
+        decoder_layer: an instance of the TransformerDecoderLayer() class (required). 
+        num_layers: the number of sub-decoder-layers in the decoder (required).
+        tgt_embed: the embedding model for the target sequence (optional).
+        pos_encoder: the positional encoding model for the target sequence (optional).
+        norm: the layer normalization component (optional).
+
+    Examples::
+        >>> transformer_decoder = nn.TransformerDecoder(decoder_layer, num_layers)
+    """
+
+    def __init__(self, decoder_layer, num_layers, tgt_embed=None, pos_encoder=None, norm=None):
+        super(TransformerDecoder, self).__init__()
+        self.layers = _get_clones(decoder_layer, num_layers)
+        self.num_layers = num_layers
+        self.tgt_embed = tgt_embed
+        self.pos_encoder = pos_encoder
+        self.norm = norm
+
+    def forward(self, tgt, memory, tgt_mask=None, memory_mask=None):
+        r"""Pass the inputs (and mask) through the decoder layer in turn.
+
+        Args:
+            tgt: the sequence to the decoder (required).
+            memory: the sequnce from the last layer of the encoder (required).
+            tgt_mask: the mask for the tgt sequence (optional).
+            memory_mask: the mask for the memory sequence (optional).
+
+        Shape:
+            see the docs in Transformer class.
+        """
+        output = tgt
+        if self.tgt_embed:
+            output = self.tgt_embed(output)
+
+        if self.pos_encoder:
+            output = self.pos_encoder(output)
+
+        for i in range(self.num_layers):
+            output = self.layers[i](output, memory, tgt_mask, memory_mask)
+
+        if self.norm:
+            output = self.norm(output)
+
+        return output
+
+class TransformerEncoderLayer(Module):
+    r"""TransformerEncoderLayer is made up of self-attn and feedforward network.
+
+    Args:
+        d_model: the embed dim (required).
+        nhead: the number of heads in the multiheadattention models (required).
+        d_ff: the dimension of the feedforward network model (default=2048).
+        dropout: the dropout value (default=0.1).
+
+    Examples::
+        >>> encoder_layer = nn.TransformerEncoderLayer(d_model, nhead)
+    """
+
+    def __init__(self, d_model, nhead, d_ff=2048, dropout=0.1):
+        super(TransformerEncoderLayer, self).__init__()
+        self.self_attn = MultiheadAttention(d_model, nhead, dropout=dropout)
+        self.ff = FeedForward(d_model, d_ff=d_ff, dropout=dropout)
```

Comment:
I suggest we inline the implementation of FeedForward until we're implementing a fused layer or such.

Reply:
The reasoning is that FeedForward is quite small and we've historically decided to not introduce an abstraction for this type of model, but rather require the user to implement it herself.

Reply:
When you inline it, do mark it as such, so that we can possibly pull it back out later on

Reply:
yea I agree.